### DISCUSSION TITLE
Some question with regards to recent changes

### DISCUSSION THREAD

**Discussion by PeterKottas** [react-page/react-page#989] [General] [answered]
Hey @macrozone, long time no see! Hope all is well. Kudos to the recent changes. Lots of good work you did there mate. Couple pointers/questions though.

1. Removal of the Component from Cell Plugin.
The component was a place where you could create an API that would be reused between Render and Controls. A slightly simplistic example would be the spacer. We have the react-resizable in the render and you can have a slider in controls (yes, now in code, the controls are generated from schema but bear with me). Both share the same "Plugin API" yet now you have to reimplement. Yes in this case it's simple, but imagine this can be complex.

2. The uiTranslatorFunction. 
Naturally, this is inspired by the i18next react implementation. The only problem I see is now we have half of the stuff using an object with all possible translations while some other stuff uses this new function. I wonder if we are not binding people's hands too much with the function. I am not an expert on all different i18n packages but some might be hard to implement using the function. Also, it's hard to know what keys to expect and it can change silently while with an object you have the type-safety. Most libs I've seen use the "object approach". I don't have a strong preference but I don't like the lack of type-safety with the function approach. I wonder what your take is on this. 

3. Some of the renames seem a little bit arbitrary. E.g. changing the "state" to "data", "name" to "id", "text" to "title" etc. Especially for people with a lot of custom plugins ( hi :p ) this is a huge pain. I think you kept the API unbelievably similar taking into consideration how much you changed on the inside (love the hooks and a lot of other stuff), but some of these were maybe let's say avoidable :D No big deal, but next time we should try to limit these IMHO.

**Comment by macrozone** [comment, marked as answer]
hi @PeterKottas , good to see you back!


1. ) There is very little shared code in Render and Controls, In fact, its discouraged to do so as you probably bloat your Component in the render and therefore the bundlesize for "read only" visitors. Not sure if in understand the spacer example as it does not share code between renderer and controls. Having a single component for both was an antipattern and therefore removed. Sharing code is still no problem at all as you are free to import whatever you need. A trickier example is `slate` by the way as Renderer and Controls need indeed to share a slate-instance. Therefore i introduced `Provider` which you can use to pass shared context to both Renderer and Controls. Its currently the only example i know that requires this kind of shared instance.

2.) I am also not 100% happy with it, but i did not use it. We had a lengthy discussion here: https://github.com/react-page/react-page/pull/935 and agreed that its a start. Maybe @rexebin can share some thoughts about your concerns and how we can move forward with ui-translations

3.) Totally understand you there and i had the same thoughts, but i decided to do these changes nevertheless. The properties were first deprecated and than removed, so there was some grace-period when both variants worked ;-) I also had to migrate some dozens of plugins, but it was relativly easy to do (with typescript, i don't know, 30min?). Some thoughts why i did it:

`state` --> `data`: "state" is kindof overloaded in react and people could wrongly interpret it to being local (non-persisted) state. Also slate uses `data` to store properties on nodes, so it made sense to rename it

`name` --> `id` and `text` --> `title`: these two properties confused myself a multiple times. `id` was chosen to highlight that this has to be unique and also because `name` looked like to be some "display name". `text` was also not very meaningful.

My goal was totally to not break too many public api (as i would myself be a victim of that ;-) ), but to avoid breaking in the future so I deciced to break now (for the 1.0.0 release). (custom) CellPlugins are such an important part of the editor in my opinion, i wanted to make them as straight-forward and clear as possible.

From 1.0.0 on breaking changes will always result in a major version bumb (if commits are annotated correctly). We should not break mindlessly, but we also should not fear to break. Its important to communicate migrations correctly.

For the CellPlugin changes, you could also easily create a migrate function and wrap your plugins with it.

**Reply by PeterKottas** [reply]
Cheers man!

I needed to spend some quality alone time with the code before I replied :D 

1. Now I understand what/why/how you did and I think I am fine with it. However, the main motivation came a little bit from misunderstanding IMHO. I think I should have focused more on documenting this and thus I could have avoided most of this work :D Well, I am sloppy. Anyways, just for your understanding:
Before you could easily choose not to include controls in the component. You simply used the createPlugin function without the controls. The component rendered the controls optionally therefore the package size would have stayed small. 

Let me now make a full circle and try to explain one more thing. The way it's done now, I argue that it's not really split into Controls and Renderer. It is in fact split into "InEditor" component and "InBottomToolbar" component. Why do I believe this is important (risking sounding like a complete douchebag :D)? It became clear, that a number of components are "controlled" not only from the bottom toolbar but also from the "InEditor" portion. That is I think a bit of a confusing point that results in the following. [New renderer code](https://github.com/react-page/react-page/blob/master/packages/plugins/content/spacer/src/Renderer/SpacerHtmlRenderer.tsx) vs [old renderer code](https://github.com/react-page/react-page/blob/v0.30.0/packages/plugins/content/spacer/src/Renderer/SpacerHtmlRenderer.tsx). You can see that in the old days, the renderer was truly a renderer, but now, you are forced to use the SpacerResizable, which is clearly a control component (yes, it's loaded lazily but so were the controls before). That means there is now no component (in many places) that truly serves only the purpose of rendering.

Luckily it's not all gloomy :D What I didn't like before was the plugin API was always there (in the component). Now that is moved in the provider that is rendered optionally which is a great touch! I spent a lot of time refactoring around this yesterday with some great results. One thing I would say though is:

It should be made easy and clean (maybe even automated via property) for people to know if the provider is needed in the readOnly editor or not.  

One quick note about the API as I feel that maybe that is another part that was a little bit misunderstood. The main way why/how I share API between render and controls is Preview functionality. I see you removed this capability from the Background plugin (I think maybe unintentionally). The way this works is you can change state, e.g. resize spacer or change background color, and have instant (instant because you don't have to use onChange and thus don't mess up the undo/redo state and performance) feedback in the editor. Only once you release (slider, color picker ...) then it's committed. I do this in almost every plugin as it provides a very good user experience. I think this was removed from almost every plugin so I thought maybe it was unintentional or just an oversight. 

Overall, I think you solved one issue while introducing one problem :D Isn't coding just beautiful?

I believe I might see a way to improve this but it needs your feedback. What if ... :
The Controls would render directly into the editor (like before) but we would expose a UI Component "BottomToolbarPortal" that would "Portal" your controls intended for the bottom toolbar into the correct place while giving you the freedom to have the in editor controls and bottom controls in one place. We could also have a property on this component that would identify wherein the bottom toolbar you want this to appear. What do you think? (I personally think this might be cool, but maybe I am missing something)

2. Agreed, I don't like it either. I think this should be removed TBH. The natural way of working with i18n is:
a) Start using the lib
b) Realize I need i18n
c) Look at props, find translations
d) Look at translations and provide mine based on default values
e) All props in the translation interface are required, so if anything changes in the future version, I will be alerted.
Now c) and d) goes like ... hmm there is this function ... What should I do with it, what are the keys. And what if they later silently add new strings? My app will become broken and I won't even know. While e) doesn't exist at all. Don't like this one bit unless I am missing something major.

3. No worries, I've been where you were. It's always tempting. One hard thing in coding after is naming things :) It's certainly not too hard to change, (look at how often the do this in slate :D )

**Reply by PeterKottas** [reply]
3. Not to mention stuff like [this](https://github.com/rexebin/react-page/blob/5f2466b686bf866ad291ab52671370ef92e5171d/packages/editor/src/ui/I18nTools/I18nDialog.tsx#L76). Any person who ever worked with i18n would laugh you from the room if they've seen this :D

**Reply by PeterKottas** [reply]
Just started working on Slate ... man this portaling described in 2. is IMHO very much a necessity.

**Reply by macrozone** [reply]
can only give short feedback:

> That is I think a bit of a confusing point that results in the following. New renderer code vs old renderer code. You can see that in the old days, the renderer was truly a renderer, but now, you are forced to use the SpacerResizable, which is clearly a control component (yes, it's loaded lazily but so were the controls before). That means there is now no component (in many places) that truly serves only the purpose of rendering.

Ok, i see the point with the spacer, but i strongly believe that this is an exception. Being able to drag the spacer around is just a nice gimmick (and apart from slate the only plugin that has this sort of interaction directly in the renderer). It would still work without that and many editors would probably prefer to have just a form where they can enter precise values.

There are two core reasons why i strongly believe this new abstraction is better:

- I think showing the actual content how it is for "visitors" (non-editors) is more important, in particular if you also consider to render the content in other environments (e.g. react-native or a pdf)
- developers that use the editor want enable their customers to add any content they want. This content is usually a Component with some simple props. So my goal was to make that as streamlined as possible. I achieved this with two things: (1) be able to define a configuration for the plugin without any boilerplate (using schema-driven forms). (2) a clear and simple api that is easy to understand and which makes it clear what will be visible to visitors and what to editors

additionally DRY-principle (don't repeat yourself) clearly demanded this approach (in 90% Component was `readOnly ? Renderer : Controls`). Developers should not be force to repeat this boilderplate.

In my projects, nearly all plugins (Except slate) are custom, i use no of the plugins we provide. (I even sometimes think we should remove these from the repo). I want to have an easy time adding a Component as a new CellPlugin (just look at the nightmareish code of the background plugin's control ;-) ) 




> One quick note about the API as I feel that maybe that is another part that was a little bit misunderstood. The main way why/how I share API between render and controls is Preview functionality.

you can do that. Simply call `onChange`, no need to have internal state. We optimized a lot so that it usually is no problem if you send a lot of `onChange` events in a short time. (The only exception is maybe an image that you don't upload right away)

>  I see you removed this capability from the Background plugin (I think maybe unintentionally).

probably ;-) but that thing is a mess anyway (altough it looks nice), I never used it apart from the example project

> The Controls would render directly into the editor (like before) but we would expose a UI Component "BottomToolbarPortal" that would "Portal" your controls intended for the bottom toolbar into the correct place while giving you the freedom to have the in editor controls and bottom controls in one place. We could also have a property on this component that would identify wherein the bottom toolbar you want this to appear. What do you think? (I personally think this might be cool, but maybe I am missing something)

As argumented above, i think rendering controls directly into the editor is often not what you want. I am against making the exception "easy" and the default "hard". 
And i mean, you can already do that if you want. `CellPluginRenderer` takes all properties that are also passed to controls, so you can call `onChange` from the renderer if you want (like in the spacer plugin)


> Overall, I think you solved one issue while introducing one problem :D Isn't coding just beautiful?

yes, it is ;-) i don't think i introduced any problem. You can prove me wrong by showing a specific use case you can't solve

**Reply by PeterKottas** [reply]
It was always hard to quantify valid arguments about this repo. E.g. 

> Ok, I see the point with the spacer, but I strongly believe that this is an exception

This is actually every single one of my plugins :D

> probably ;-) but that thing is a mess anyway (although it looks nice), I never used it apart from the example project

Most commonly used plugin for us ...

> you can do that. Simply call onChange, no need to have internal state. We optimized a lot so that it usually is no problem if you send a lot of onChange events in a short time.

Tried, but it's terrible. Especially with a big document. Not to mention on a tablet or slow pc. Also, I believe it breaks the undo/redo, no? I think there is some throttle in it. But still it floods it with BS after a little while.

So taking that into consideration it almost feels like it's impossible to find common ground. Normally, you would have feedback from users, but I don't think there are many people who use this in a serious manner (or as serious as the 2 of us may be).

All that being said, I believe I can achieve everything I mentioned in my own code. E.g. provide custom BottomToolbar with portals and render everything insider the renderer (ignoring controls), split internally and render into bottom toolbar using portals. 

> yes, it is ;-) i don't think i introduced any problem. You can prove me wrong by showing a specific use case you can't solve

Maybe just in terms of code migration. But it's all good. I can work it out on our end. Or to be more precise, it is a problem for the way we handle plugins, but who knows, maybe we are the exception, maybe you are. No way to know.

Inclined towards don't break what works approach :) But the 2. stands with the i18n. That approach imho is just not good.

**Reply by macrozone** [reply]
feel free to push a pull request, but please show a precice and simple use case in the examples! I think that is the best way to argument about api design.

I would recommend to see if you find a common pattern, wrap that into a component or function and then try to find a place to integrate it into the editor.

I won't block any PR just because i won't use it (or don't like it, its not my repo after all), but it should be clear how and when to use this feature. 

I also think my point of view is from a developer perspective who wants to add this editor to many different projects without any hazzle and boiler plate (i just can't afford maintaing much technical debt), whereas you probably have a one bigger project with special needs if i remember correctly.  So its important to narrow down these special needs to something simple and clean.

**Reply by macrozone** [reply]
(i am away for 2.5 weeks, so I won't be able to answer much 😺 )

**Reply by PeterKottas** [reply]
I'll see what I am left with once I am done with the upgrade and if there's anything noteworthy to contribute. While the API seems like a bit of a mess in general, I do believe it's extremely flexible so you can pretty much achieve anything ATM. I definitely don't want to make it grow unless it's truly necessary. 

> I also think my point of view is from a developer perspective who wants to add this editor to many different projects without any hazzle and boiler plate (i just can't afford maintaining much technical debt), whereas you probably have a one bigger project with special needs if i remember correctly.

That's a very good point. Naturally taking that into consideration, we approach things differently. E.g. we can afford to spend more time on fancy little bits like instant preview animations and stuff like that. While for you, that would be a mess to maintain. 

Worry not, enjoy your time off mate! I think I now understand quite a bit behind the motivation for recent changes etc. Cheers for that!

**Reply by macrozone** [reply]
> While the API seems like a bit of a mess in general, 

Please try to give constructive feedback and provide facts for your argument. Nothing and nobody is perfect. Personally i believe its overall much much better than it was 0.20-something 😉

> Worry not, enjoy your time off mate! I think I now understand quite a bit behind the motivation for recent changes etc. Cheers for that!

thank you, thats good to hear! Will do that! (without many screen-time hopefully...)

**Reply by PeterKottas** [reply]
> Please try to give constructive feedback and provide facts for your argument. Nothing and nobody is perfect. Personally i believe its overall much much better than it was 0.20-something 😉

I meant in general. Especially on the inside. It certainly improved a lot!
